src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -42,6 +42,31 @@
 
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  // PROTO Data structure:
+  // Parse the contents of graph.txt into a proto Graph object, and extract
+  // information
+  // from the proto object into a map. This is used to store the proto input and
+  // isn't updated
+  // with mutations.
+  private Graph protoGraph = null;
+  private Map<String, Node> protoNodesMap = null;
+
+  // GRAPH Data structures:
+  // Create an undirected graph data structure to store the information, and
+  // map each node name in the graph to the GraphNode objects. This is the graph &
+  // map
+  // we update with mutations
+  private MutableGraph<GraphNode> graph = null;
+  private HashMap<String, GraphNode> graphNodesMap = null;
+
+  private ImmutableGraph<GraphNode> graphOriginal = null; // never undergoes mutations
+
+  // Data structure that stores the roots of the graph across mutations
+  // Roots are nodes with no in-edges
+  private HashSet<String> roots = null;
+
+  private List<Mutation> mutList = null;
```

Comment:
I feel like this could be pulled into a separate data class? :)

Reply:
Do you mean a data class with roots and a list of mutations as its fields?

Reply:
I mean all the protoGraph, protoNodesMap, graph, graphNodesMap, graphOriginal, roots, mutList could be in a separate data class (or maybe two classes, one for graph and one for mutations).

This decouples the "data class" and the servlet, even if we still need to construct the data class within the get method, it is easy to change that in the future. Does this make sense?

Reply:
yea that makes sense!